File: leetcode/All/NumberOfIslands.java
import java.io.*;
import java.util.*;

class Pair {
    public int f, s;
    Pair(int x, int y) {
        f=x;
        s=y;
    }
}
class NumberOfIslands {

    // Implement your solution by completing the below function

    
    void dfs(Pair src, int n, int m, char[][] grid, boolean[][] vis) {
        vis[src.f][src.s] = true;

        if(src.f-1>=0 && !vis[src.f-1][src.s] && grid[src.f-1][src.s]=='1')
            dfs(new Pair(src.f-1, src.s), n, m, grid, vis);
        if(src.f+1<n && !vis[src.f+1][src.s] && grid[src.f+1][src.s]=='1')
            dfs(new Pair(src.f+1, src.s), n, m, grid, vis);
        if(src.s-1>=0 && !vis[src.f][src.s-1] && grid[src.f][src.s-1]=='1')
            dfs(new Pair(src.f, src.s-1), n, m, grid, vis);
        if(src.s+1<m && !vis[src.f][src.s+1] && grid[src.f][src.s+1]=='1')
            dfs(new Pair(src.f, src.s+1), n, m, grid, vis);
    }
    public int numIslands(char[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int cnt=0;
        boolean[][] vis = new boolean[n][m];
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                vis[i][j]=false;
            }
        }
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(!vis[i][j] && grid[i][j]=='1') {
                    dfs(new Pair(i, j), n, m, grid, vis);
                    cnt++;
                }
            }
        }
        return cnt;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int rows = scanner.nextInt();
        int columns = scanner.nextInt();
        if(rows==0 || columns==0) {
            System.out.println(0);
            scanner.close();
            return;
        }
        char[][] grid = new char[rows][columns];

        for (int i = 0; i < rows; ++i) {
            String s = scanner.next();
            for (int j = 0; j < columns; ++j) {
                grid[i][j] = s.charAt(j);
            }
        }
        scanner.close();
        int result = new NumberOfIslands().numIslands(grid);
        System.out.println(result);
    }
}

File: leetcode/All/Number of Islands.cpp
class Solution {
public:
    int d[4][2]={{0, -1}, {0, 1}, {1, 0}, {-1, 0}};
    void dfs(int i, int j, int n, int m, vector<vector<char>>& grid, vector<vector<bool>>& vis)     {
        vis[i][j]=true;
        for(int k=0; k<4; k++) {
            int ni=i+d[k][0], nj=j+d[k][1];
            if(ni>-1 and ni<n and nj>-1 and nj<m and !vis[ni][nj] and (grid[ni][nj]-'0'))
                dfs(ni, nj, n, m, grid, vis);
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        if(grid.size()==0) return 0;
        int n=grid.size(), m=grid[0].size(), island=0;
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        for(int i=0; i<n; i++)
            for(int j=0; j<m; j++)
                if((grid[i][j]-'0') and !vis[i][j])
                {
                    dfs(i, j, n, m, grid, vis);
                    // cout<<i<<' '<<j<<endl;
                    island++;
                }
        return island;
    }
};